bgpd/bgp_aspath.c

```diff
@@ -1316,7 +1318,9 @@ struct aspath *aspath_remove_private_asns(struct aspath *aspath, as_t peer_asn)
 		// there are public ASNs then come back and fill in only the
 		// public ASNs.
 		else {
-			new_seg = assegment_new(seg->type, public);
+			/* length needs to account for all retained ASNs
+			 * (public or peer_asn), not just public */
+			new_seg = assegment_new(seg->type, (public + peer));
```

Comment:
No, we're allocating a new path segment to hold the modified path, which needs to be big enough to account for the total number of ASNs we plan to retain, i.e. every instance of a public ASN + every instance of a private ASN that matches the peer's.

Since before this codepath didn't account for the peer's ASN, the length used to be just `public`.  Now since we're also retaining all instances of the peer's ASN, `(public + peer)` is the length we use (if the peer's ASN is public, that is accounted once in `public`).

Reply:
So, if the AS-PATH comes from a peer (peer_asn: 65001) `65001 65002 65003 65001 65003 65001`, we _leave_ AS-PATH to `65001 65001 65001` here?

Reply:
Correct.  The local ASN is prepended in a later part of update tx processing.

Reply:
I'm just curious what the correct approach should be. To keep 65001 65001 65001 or just a single 65001. But otherwise, it could break as-path pretending...

Reply:
The primary motivation for the change is to trigger the route to be dropped by loop prevention.  Unless the peer is configured with `allowas-in` AND the peer's ASN is in the path, the actual path length wouldn't really come into play.

Even if we hit those conditions, my gut reaction is that retaining each instance would be better for "preserving" some of the AS-Path length.

I'm not opposed to other views on what the "correct" approach would be, but I would mention that the existing logic isn't being modified in this PR -- just expanded to include `remove-private-AS` configs that don't have the `replace-AS` keyword.
I think it might be cleaner to split out any changes to that logic into its own PR (which I'm also happy to write, should we decide is worthwhile).  What do you think?

Reply:
👍